### PR TITLE
introduce general-purpose service dependency graph traversal functions

### PR DESCRIPTION
**Pull request by ndeloof**
This introduces in compose-go the `InDependencyOrder` utility method which was designed for docker/compose
Comes with a revisited (simplified) implementation, + a new `CollectInDependencyOrder` utility to run function on services but also collect results into a `map[string]T`

Also revisit `WithServices` (same purpose, but not running in dependency order neither concurrently) to pass service name to function, and also adopt functional options pattern to declare dependency policy - could be used in future to introduce more options

### CODE REVIEW

File: utils/set.go
Comment (glours) [inline utils/set.go:1]: And why not an immutable Set? 😄

File: utils/collectionutils.go [hunk 1]
```diff
@@ -16,13 +16,17 @@
...
+)
 
-func MapKeys[T comparable, U any](theMap map[T]U) []T {
-	var result []T
+func MapKeys[T constraints.Ordered, U any](theMap map[T]U) []T {
+	result := make([]T, 0, len(theMap))
 	for key := range theMap {
 		result = append(result, key)
 	}
+	slices.Sort(result)
```
Comment (glours) [inline utils/collectionutils.go:29]: Now we can use https://pkg.go.dev/golang.org/x/exp/maps#Keys
If we want to keep it sorted: 
```suggestion
func MapKeys[T constraints.Ordered, U any](theMap map[T]U) []T {
	result := slices.Sort(maps.Keys(theMap))
```

**Review by glours — APPROVED** [review]
LGTM

File: utils/set.go
Comment (ndeloof) [reply utils/set.go:1]: this is the code we had in docker/compose, but I'm fine to change it to another implementation

**Pull request merged**